src/node_api.cc

```diff
@@ -0,0 +1,2430 @@
+    U(v8::Local<v8::Value> _l) : l(_l) {}
+  } u(local);
+  assert(sizeof(u.v) == sizeof(u.l));
+  return u.v;
+}
+
+v8::Local<v8::Value> V8LocalValueFromJsValue(napi_value v) {
+  // Likewise awkward
+  union U {
+    napi_value v;
+    v8::Local<v8::Value> l;
+    U(napi_value _v) : v(_v) {}
+  } u(v);
+  assert(sizeof(u.v) == sizeof(u.l));
+  return u.l;
+}
+
+static v8::Local<v8::Function> V8LocalFunctionFromJsValue(napi_value v) {
+  // Likewise awkward
+  union U {
+    napi_value v;
+    v8::Local<v8::Function> f;
+    U(napi_value _v) : v(_v) {}
+  } u(v);
+  assert(sizeof(u.v) == sizeof(u.f));
+  return u.f;
+}
+
+// Wrapper around v8::Persistent that implements reference counting.
+class Reference {
+ public:
+  Reference(v8::Isolate* isolate,
+            v8::Local<v8::Value> value,
+            int initialRefcount,
+            bool deleteSelf,
+            napi_finalize finalizeCallback = nullptr,
+            void* finalizeData = nullptr,
+            void* finalizeHint = nullptr)
+       : _isolate(isolate),
+        _persistent(isolate, value),
+        _refcount(initialRefcount),
+        _deleteSelf(deleteSelf),
+        _finalizeCallback(finalizeCallback),
+        _finalizeData(finalizeData),
+        _finalizeHint(finalizeHint) {
+    if (initialRefcount == 0) {
+      _persistent.SetWeak(
+          this, FinalizeCallback, v8::WeakCallbackType::kParameter);
+      _persistent.MarkIndependent();
+    }
+  }
+
+  ~Reference() {
+    // The V8 Persistent class currently does not reset in its destructor:
+    // see NonCopyablePersistentTraits::kResetInDestructor = false.
+    // (Comments there claim that might change in the future.)
+    // To avoid memory leaks, it is better to reset at this time, however
+    // care must be taken to avoid attempting this after the Isolate has
+    // shut down, for example via a static (atexit) destructor.
+    _persistent.Reset();
+  }
+
+  int AddRef() {
+    if (++_refcount == 1) {
+      _persistent.ClearWeak();
+    }
+
+    return _refcount;
+  }
+
+  int Release() {
+    if (--_refcount == 0) {
+      _persistent.SetWeak(
+          this, FinalizeCallback, v8::WeakCallbackType::kParameter);
+      _persistent.MarkIndependent();
+    }
+
+    return _refcount;
+  }
+
+  v8::Local<v8::Value> Get() {
+    if (_persistent.IsEmpty()) {
+      return v8::Local<v8::Value>();
+    } else {
+      return v8::Local<v8::Value>::New(_isolate, _persistent);
+    }
+  }
+
+ private:
+  static void FinalizeCallback(const v8::WeakCallbackInfo<Reference>& data) {
+    Reference* reference = data.GetParameter();
+    reference->_persistent.Reset();
+
+    // Check before calling the finalize callback, because the callback might
+    // delete it.
+    bool deleteSelf = reference->_deleteSelf;
+
+    if (reference->_finalizeCallback != nullptr) {
+      reference->_finalizeCallback(reference->_finalizeData,
+          reference->_finalizeHint);
+    }
+
+    if (deleteSelf) {
+      delete reference;
```

Comment:
It feels counter intuitive to go to such lengths to make the public API be C, but then use `delete` on a user-supplied pointer.